### ISSUE TITLE
[Bug][ROCm/gfx942]: GLM-5.2-FP8 — first request after GPU idle emits garbage; piecewise CUDA graph cold replay corrupts the request's own prefill (workaround: cudagraph_mode=FULL_DECODE_ONLY)

### ISSUE THREAD

**Issue by zzw09773** [vllm-project/vllm#52150]
## Environment

- 8× AMD Instinct MI325X (gfx942), ROCm 7.14.0, amdgpu 6.19.14
- Images: `vllm/vllm-openai-rocm:nightly` 2026-08-12 (v0.26.1rc1.dev668+g3ee2df303) **and** 2026-08-13 (`8f515d1d`) — both reproduce
- Model: `zai-org/GLM-5.2-FP8`, TP=8, `--kv-cache-dtype auto`, `--max-num-batched-tokens 4096`, MoE backend AITER blockscale-FP8 (`fmoe_bf16_a16_blockscaleFp8...`)
- Reproduces identically at `--max-model-len` 131072 and 1048576

## 🐛 Describe the bug

After the engine sits idle for ≥ ~20–50 s, the **next** request returns garbage from its very first sampled token, e.g. `'!0.1.5,0.1 and1.1.1.0 21.1.'` (token-0/digit noise). Requests sent immediately afterwards are correct. Reproduction probability rises with idle time (~50–100 % at ≥ 50 s idle; back-to-back requests never fail, dozens of samples).

Key observations:

1. **The corrupt request never self-heals**: with `max_tokens=300` the output is noise from token 1 through token 300, while a request issued ~0.5 s later is clean — i.e. the corrupt request's *own prefill* wrote garbage into its KV/indexer cache, and every decode step then consumes that poisoned context.
2. `temperature=0`; prompts are ~6–10 tokens (below the 16-token block size, so prefix caching is not involved).
3. No errors in the server log, container healthy, RAS/ECC counters unchanged.
4. Not JIT/autotune related: Triton JIT warnings (`_cp_gather_indexer_quant_cache_kernel`, `_fp8_mqa_logits_kernel`) appear only on the very first request after startup; later corrupt events have no JIT activity.
5. A 25 s-interval keepalive (unique prompts forcing real prefills, all HTTP 200) does **not** reliably prevent it.

## Bisection

| config | result |
|---|---|
| defaults (FULL_AND_PIECEWISE graphs), 131k & 1M, nightly 08-12 & 08-13 | ❌ corrupt |
| `--enforce-eager` | ✅ clean 8/8 (startup burst + 60 s idle ×3 + 25 s idle ×2) |
| `--compilation-config='{"cudagraph_mode":"FULL_DECODE_ONLY"}'` | ✅ clean 8/8 at 1M |

→ The corruption tracks **piecewise CUDA/HIP graph replay in the prefill path** after a cold window. With prefill graph-free (eager or `FULL_DECODE_ONLY`) the model is fully healthy: fact probes pass, tool calling OK, needle-in-haystack 3/3 at 2k/6k/12k/100k/900k tokens, and a 5-minute cold soak followed by a fresh request is correct.

Other models on the same host/stack (Kimi-K3 with eager prefill + FULL decode graphs, DeepSeek-V4-Flash-0731) do not show this — consistent with the piecewise-prefill-graph attribution: decode-side graphs essentially never replay cold, because the preceding prefill has just warmed the GPU.

This smells like a ROCm/HIP graph runtime issue (queue/doorbell state after idle → first graph replay misfires) rather than a model bug, but filing here since the practical impact is "GlmMoeDsa default config silently corrupts on gfx942" and the workaround is a vLLM flag.

Related (same class — HIP graph replay silently wrong on AMD — but none match the idle-triggered gfx942 manifestation):
- ROCm/hip#3887 — HIP graph replay reads stale device-pointer values (gfx1201/RDNA4, ~200+ nodes)
- ROCm/aiter#3632 — W4A4 fused MoE kernel not HIP-graph-safe (gfx950)
- pytorch/pytorch#155684 — HIP graph misuse yields silent wrong results where CUDA raises

## Workaround

`--compilation-config='{"cudagraph_mode":"FULL_DECODE_ONLY"}'` — no measurable decode-speed loss (decode keeps FULL graphs; only prefill runs graph-free). Single-stream decode on this host: 44.5 tok/s (prose) / 61.9 tok/s (code) at 1M `max_model_len`.

## Repro

1. Serve `zai-org/GLM-5.2-FP8` on 8×gfx942 with the default compilation config.
2. `sleep 60`
3. ```bash
   curl -s localhost:8000/v1/completions -H 'content-type: application/json' \
     -d '{"model":"glm","prompt":"The capital of France is","max_tokens":16,"temperature":0}'
   ```
   → garbage (`!0.1...`). Repeat immediately → ` Paris.` (correct). Vary the prompt each round to rule out prefix-cache effects; alternate idle/burst to see the pattern.

**Comment by indivats** [comment]
@hongxiayang will look into this, could you please assign this issue to me?

**Comment by indivats** [comment]
Thanks @zzw09773 I drove this down to a specific kernel with single-variable experiments. 

The root cause appears to be that AITER's **1-stage** fused all-reduce + RMSNorm kernel (`rocm_aiter_fused_allreduce_rmsnorm` / `custom_fused_ar_rms`, `use_1stage=True`) is not HIP-graph cold-replay-safe on gfx942. The 2-stage path is clean.

On GLM-5.2-FP8, TP=8, 8x MI300X, 75 s idle, temp=0:

| Config | Cold garbage |
|---|---|
| baseline | 5/12 |
| `pass_config.fuse_allreduce_rms=false` | 0/12 |
| `use_1stage=False` (one-line force, identical graph) | 0/12 |

The third row holds the entire captured graph fixed and flips only the AR variant, which isolates the 1-stage path.

**Workaround** (keeps all cudagraphs, including piecewise prefill, so no prefill perf loss unlike `FULL_DECODE_ONLY`): --compilation-config '{"pass_config":{"fuse_allreduce_rms":false}}'
This disables only the AR+RMSNorm fusion; AITER custom AR, GEMM and MoE stay on. `VLLM_ROCM_USE_AITER_CUSTOM_AR=0` also works (disables AITER's AR entirely).

Please verify and use the workaround if needed. Filed the issue upstream at ROCm/aiter: https://github.com/ROCm/aiter/issues/4888
